File: app/robust/has_subtree.py
# coding=utf-8
# author: Xiguang Liu<g10guang@foxmail.com>
# 2018-04-27 17:44
# 题目描述：https://www.nowcoder.com/practice/6e196c44c7004d15b1610b9afca8bd88?tpId=13&tqId=11170&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    def HasSubtree(self, A, B):
        """
           用于防止 B 树一开始就是空
        """
        if B is None:
            return False
        return self.logic(A, B)

    def logic(self, A, B):
        if B is None:
            return True
        if A is None:
            return False
        r = False
        if A.val == B.val:
            r = self.logic(A.left, B.left) and self.logic(A.right, B.right)
        if not r:
            r = self.logic(A.left, B) or self.logic(A.right, B)
        return r


import unittest


class TestCase(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test_1(self):
        A = TreeNode(8)
        A.left = TreeNode(8)
        A.right = TreeNode(7)
        A.left.left = TreeNode(9)
        A.left.right = TreeNode(2)
        A.right.left = TreeNode(4)
        A.right.right = TreeNode(7)

        B = TreeNode(8)
        B.left = TreeNode(9)
        B.right = TreeNode(2)

        r = self.s.HasSubtree(A, B)

        self.assertTrue(r)

File: app/robust/has_subtree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
 *
 * @param pRoot1 TreeNode类
 * @param pRoot2 TreeNode类
 * @return bool布尔型
 */
// https://www.nowcoder.com/practice/6e196c44c7004d15b1610b9afca8bd88?tpId=13&tqId=11170&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking
func HasSubtree(pRoot1 *TreeNode, pRoot2 *TreeNode) bool {
	if pRoot2 == nil {
		return false
	}
	if pRoot1 == nil {
		return false
	}
	if isMatchSubTree(pRoot1, pRoot2) {
		return true
	}
	return HasSubtree(pRoot1.Left, pRoot2) || HasSubtree(pRoot1.Right, pRoot2)
}

func isMatchSubTree(p, q *TreeNode) bool {
	if q == nil {
		return true
	}
	if p == nil {
		return false
	}
	if p.Val == q.Val {
		return isMatchSubTree(p.Left, q.Left) && isMatchSubTree(p.Right, q.Right)
	}
	return false
}
